src/gui/dialogs/help_browser.cpp

```diff
@@ -19,82 +19,192 @@
+	initial_node.select_node(true);
+
+	on_topic_select();
+}
+
+void help_browser::add_topics_for_section(const help::section& parent_section, tree_view_node& parent_node)
+{
+	for(const help::section& section : parent_section.sections) {
+		tree_view_node& section_node = add_topic(section.id, section.title, true, parent_node);
+
+		add_topics_for_section(section, section_node);
+	}
+
+	for(const help::topic& topic : parent_section.topics) {
+		if(topic.id.compare(0, 2, "..") != 0) {
+			add_topic(topic.id, topic.title, false, parent_node);
+		}
+	}
+}
+
+tree_view_node& help_browser::add_topic(const std::string& topic_id, const std::string& topic_title,
+		bool expands, tree_view_node& parent)
+{
+	widget_data data;
+	widget_item item;
+
+	item["label"] = topic_title;
+	data.emplace("topic_name", item);
+
+	tree_view_node& new_node = parent.add_child(expands ? "section" : "topic", data);
+	new_node.set_id(std::string(expands ? "+" : "-") + topic_id);
+
+	return new_node;
+}
+void help_browser::show_topic(std::string topic_id, bool add_to_history)
+{
+	if(topic_id.empty()) {
+		return;
+	}
+
+	if(topic_id[0] == '+') {
+		topic_id.replace(topic_id.begin(), topic_id.begin() + 1, 2, '.');
+	}
+
+	if(topic_id[0] == '-') {
+		topic_id.erase(topic_id.begin(), topic_id.begin() + 1);
+	}
+
+	auto iter = parsed_pages_.find(topic_id);
+	if(iter == parsed_pages_.end()) {
+		const help::topic* topic = help::find_topic(toplevel_, topic_id);
+		if(!topic) {
+			ERR_GUI_P << "Help browser tried to show topic with id '" << topic_id
+				  << "' but that topic could not be found." << std::endl;
+			return;
+		}
 
-	for(const auto& topic : help_cfg_.child_range("topic")) {
 		widget_data data;
 		widget_item item;
 
-		item["label"] = topic["title"];
-		data.emplace("topic_name", item);
+		item["label"] = topic->title;
+		data.emplace("topic_title", item);
 
-		topic_tree.add_node("topic", data).set_id(std::to_string(id));
+		find_widget<label>(this, "topic_title", false).set_label(topic->title);
+		find_widget<rich_label>(this, "topic_text", false).set_topic(topic);
 
-		// FIXME: maybe using a multi page isn't a good idea here... :| it causes massive lag when opening.
-		item.clear();
-		data.clear();
+		invalidate_layout();
+		scrollbar_panel& scroll = find_widget<scrollbar_panel>(this, "topic_scroll_panel", false);
+		scroll.scroll_vertical_scrollbar(scrollbar_base::BEGIN);
+	}
 
-		item["label"] = topic["text"].empty() ? "" : topic["text"].str();
-		data.emplace("topic_text", item);
+	if (add_to_history) {
+		// history pos is 0 initially, so it's already at first entry
+		// no need increment first time
+		if (!history_.empty()) {
+			history_pos_++;
+		}
+		history_.push_back(topic_id);
 
-		topic_pages.add_page(data);
+		find_widget<button>(this, "back", false).set_active(history_pos_ != 0);
 
-		++id;
 	}
+}
 
-	on_topic_select();
+void help_browser::on_link_click(std::string link)
```

Comment:
If this is just calling `show_topic`, suggest removing this function and binding `show_topic` directly

Reply:
updated.